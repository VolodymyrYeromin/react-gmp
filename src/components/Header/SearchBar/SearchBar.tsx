import styles from "./searchBar.module.scss";
import {useFormik} from "formik";
import {useRouter} from "next/router";
import {FC} from "react";
import useCustomNavigation from "../../../hooks/useCustomNavigation"

const SearchBar:FC = () => {
    const nextRouter = useRouter();
    const formik = useFormik({
        initialValues: {
            search_query: nextRouter.query.searchQuery ? nextRouter.query.searchQuery : ''
        },
        onSubmit: (values) => {
            useCustomNavigation({navigate: nextRouter.push, searchQuery: values.search_query, genre: nextRouter.query.genre, sortBy: nextRouter.query.sortBy, movie: nextRouter.query.movie});
        }
    })

    return (
        <div className={styles.searchBlock} data-testid="search-block">
            <h1>Find your movie</h1>
            <form onSubmit={formik.handleSubmit} className={styles.searchBar}>
                <input className={styles.searchQuery} name="search_query" type="text" placeholder="What do you want to watch?" value={formik.values.search_query} onChange={formik.handleChange} data-testid="search-input"/>
                <input className={styles.searchButton} type="submit" value="Search" data-testid="search-button"/>
            </form>
        </div>
    );
};

export default SearchBar;
