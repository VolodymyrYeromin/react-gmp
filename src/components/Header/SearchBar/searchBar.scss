.search-block {
  padding: 0 80px;
  margin-top: 50px;
  h1 {
    font-weight: 300;
    font-size: 40px;
    color: var(--white);
    line-height: 48px;
    text-transform: uppercase;
    margin: 0 0 50px
  }
  .search-bar {
    display: flex;
    .search-query {
      color: var(--white);
      background-color: #323232CC;
      width: 70%;
      font-size: 20px;
      line-height: 24px;
      font-family: 'Montserrat', sans-serif;
      font-weight: 400;
      padding: 14px;
      margin-right: 14px;
      max-height: 57px;
      border-radius: 4px;
      border-color: transparent;
      &:focus {
        outline: none;
      }
      &::placeholder {
        color: var(--white);
        opacity: 0.3;
      }
    }
  }
  .search-button {
    padding: 17px 73px;
    height: 59px;
    font-size: 20px;
    line-height: 24px;
    border-radius: 4px;
    cursor: pointer;
    background-color: var(--red);
    color: var(--white);
    text-transform: uppercase;
  }
}
