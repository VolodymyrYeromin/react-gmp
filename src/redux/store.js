import { configureStore } from "@reduxjs/toolkit";
import moviesReducer from "./features/movies/moviesSlice";
import modalReducer from "./features/modal/modalSlice";
import sortFilterBarReducer from "./features/sortFilterBar/sortFilterBarSlice";
import selectedMovieReducer from "./features/selectedMovie/selectedMovieSlice";


export const store = configureStore({
    reducer: {
        movies: moviesReducer,
        modal: modalReducer,
        sortFilterBar: sortFilterBarReducer,
        selectedMovie: selectedMovieReducer,
    },
});
