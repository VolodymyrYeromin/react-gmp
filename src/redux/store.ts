import { configureStore } from "@reduxjs/toolkit";
import moviesReducer from "./features/movies/moviesSlice";
import modalReducer from "./features/modal/modalSlice";
import selectedMovieReducer from "./features/selectedMovie/selectedMovieSlice";


export const store = configureStore({
    reducer: {
        movies: moviesReducer,
        modal: modalReducer,
        selectedMovie: selectedMovieReducer,
    },
});

export type AppDispatch = typeof store.dispatch;
