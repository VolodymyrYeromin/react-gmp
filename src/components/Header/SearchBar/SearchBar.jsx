import "./searchBar.scss";
import {useFormik} from "formik";
import {useDispatch} from "react-redux";
import {getMovies} from "../../../redux/features/movies/moviesSlice";
import {useNavigate, useParams} from "react-router-dom";
import useQuery from "../../../hooks/useQuery";
import constants from "../../../constants";
import useDidMountEffect from "../../../hooks/useDidMountEffect";
import useCustomNavigation from "../../../hooks/useCustomNavigation";
import {useEffect} from "react";

const SearchBar = () => {
    const {searchQuery} = useParams();
    const navigate = useNavigate();
    const query = useQuery();
    const sortBy = query.get(constants.queryParams.SORT_BY);
    const genre = query.get(constants.queryParams.GENRE);

    const dispatch = useDispatch();
    const formik = useFormik({
        initialValues: {
            search_query: ''
        },
        onSubmit: (values) => {
            useCustomNavigation({navigate, searchQuery: values.search_query, genre, sortBy});
        }
    })
    useDidMountEffect(() => {
        dispatch(getMovies({searchQuery, sorting: sortBy, filtering: genre}));
    }, [searchQuery]);

    useEffect(() => {
        formik.setFieldValue('search_query', searchQuery ? searchQuery : '', false);
    }, []);

    return (
        <div className="search-block">
            <h1>Find your movie</h1>
            <form onSubmit={formik.handleSubmit} className="search-bar">
                <input className="search-query" name="search_query" type="text" placeholder="What do you want to watch?" value={formik.values.search_query} onChange={formik.handleChange}/>
                <input className="search-button" type="submit" value="Search"/>
            </form>
        </div>
    );
};

export default SearchBar;
