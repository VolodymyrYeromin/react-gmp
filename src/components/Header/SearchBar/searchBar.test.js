import SearchBar from "./SearchBar";
import {fireEvent, render, screen, waitFor} from "@testing-library/react";
import {MemoryRouter} from "react-router-dom";
import {Provider} from "react-redux";
import {store} from "../../../redux/store";
import {act} from "react-dom/test-utils";
import userEvent from "@testing-library/user-event";

const mockedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedNavigate
}));

describe("Test search bar", () => {
    beforeEach(() => {
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <SearchBar/>
                </MemoryRouter>
            </Provider>
        )
    })

    test('search bar matching screenshot', () => {
        const searchBlock = screen.getByTestId('search-block');

        expect(searchBlock).toMatchSnapshot();
    });

    test('search bar input event and search request', async () => {
        const searchInput = screen.getByTestId('search-input');
        const searchButton = screen.getByTestId('search-button');

        await act(() => fireEvent.change(searchInput, {target: {value: 'avengers'}}));
        expect(searchInput.value).toBe('avengers');

        userEvent.click(searchButton)
        await waitFor(() => {
            expect(mockedNavigate).toHaveBeenCalledTimes(1);
        });
    })
})
