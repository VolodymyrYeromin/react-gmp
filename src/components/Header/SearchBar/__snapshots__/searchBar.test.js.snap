// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test search bar search bar matching screenshot 1`] = `
<div
  class="search-block"
  data-testid="search-block"
>
  <h1>
    Find your movie
  </h1>
  <form
    class="search-bar"
  >
    <input
      class="search-query"
      data-testid="search-input"
      name="search_query"
      placeholder="What do you want to watch?"
      type="text"
      value=""
    />
    <input
      class="search-button"
      data-testid="search-button"
      type="submit"
      value="Search"
    />
  </form>
</div>
`;
